File: workflow/kernel/tools/render_handoff.py
#!/usr/bin/env python3
"""Deterministic resume/handoff digest for an Amanar workflow.

Read-only. Renders controller state as Markdown from `.amanar/workflow.json` and
`.amanar/run/`: the derived state, each receipt's current/stale/missing verdict, an
id/artifact closet, and the ordered rebuild to `verified`. It reuses the
controller's own freshness logic rather than reimplementing it, and is deliberately
*not* part of the frozen CLI or the workflow contract.

Resume/handoff pattern reimplemented from iamneilroberts/claude-skills (MIT).

Usage: `python3 .amanar/kernel/tools/render_handoff.py [--root PATH]`
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))  # kernel dir -> amanar_workflow

from amanar_workflow import cli  # noqa: E402
from amanar_workflow.contract import load, workflow_hash  # noqa: E402
from amanar_workflow.errors import WorkflowError  # noqa: E402
from amanar_workflow.receipts import output_problem, receipt_problem, source_snapshot  # noqa: E402
from amanar_workflow.state import read as read_state  # noqa: E402


def _verdict(location: dict, contract: dict, check: dict, current_digest: str | None) -> tuple[str, str]:
    receipt_path = location["receipts"] / f"{check['id']}.json"
    if not receipt_path.is_file():
        return "MISSING", "no receipt recorded"
    if current_digest is None:
        return "UNKNOWN", "source digest unavailable (not a committed git tree)"
    receipt = cli.read_json(receipt_path)
    problem = receipt_problem(receipt, contract, check, current_digest)
    if problem is None:
        problem = output_problem(receipt, check, location["run"] / "output")
    if problem:
        return "STALE", problem
    return "CURRENT", f"passed, tests={receipt.get('discoveredTests')}, at {receipt.get('recordedAt')}"


def render(root: Path) -> str:
    location = cli.paths(root)
    contract = load(location["contract"])
    state = read_state(location["state"])
    recorded = "planned" if state is None else state["status"]
    try:
        current_digest = None if state is None else source_snapshot(root)["digest"]
    except WorkflowError:
        current_digest = None

    verdicts = [(c["id"], *_verdict(location, contract, c, current_digest)) for c in contract["checks"]]
    stale = any(tag != "CURRENT" for _, tag, _ in verdicts)
    effective = "implementing" if recorded == "verified" and stale else recorded

    out: list[str] = [f"# Workflow handoff — {contract['id']}", ""]
    label = f"**{effective}**" + (f" (recorded {recorded})" if effective != recorded else "")
    out.append(f"- State: {label}")
    out.append(f"- Objective: {contract['objective']}")
    if state and state.get("blockReason"):
        out.append(f"- Blocked: {state['blockReason']}")

    out += ["", "## Receipts"]
    for cid, tag, detail in verdicts:
        out.append(f"- `{cid}`: **{tag}** — {detail}")

    out += ["", "## Closet"]
    out.append(f"- Workflow id: `{contract['id']}`")
    out.append(f"- Workflow hash: `{workflow_hash(contract)}`")
    out.append("- Scope: " + (", ".join(f"`{p}`" for p in contract["scope"]) or "(none)"))
    if contract.get("exclusions"):
        out.append("- Exclusions: " + ", ".join(f"`{p}`" for p in contract["exclusions"]))
    out.append("- Artifacts: " + (", ".join(f"`{p}`" for p in contract["artifacts"]) or "(none)"))

    out += ["", "## Rebuild to verified"]
    if effective == "verified":
        out.append("- Already verified with current receipts. No action needed.")
    else:
        steps: list[str] = []
        if effective in ("planned", "blocked"):
            steps.append("`begin` (required before any check)")
        steps += [f"`run-check {cid}` (currently {tag})" for cid, tag, _ in verdicts if tag != "CURRENT"]
        steps.append("`verify`")
        out += [f"{i}. {step}" for i, step in enumerate(steps, 1)]
    out.append("")
    return "\n".join(out)


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(prog="render_handoff")
    parser.add_argument("--root", default=".", help="repository root containing .amanar/")
    args = parser.parse_args(argv)
    try:
        sys.stdout.write(render(Path(args.root).resolve()))
    except WorkflowError as exc:
        print(f"render_handoff error: {exc}", file=sys.stderr)
        raise SystemExit(1) from exc


if __name__ == "__main__":
    main()

File: workflow/kernel/src/tools/renderHandoff.ts
/**
 * Deterministic resume/handoff digest for an Amanar workflow.
 * Mirrors tools/render_handoff.py exactly.
 */

import fs from 'node:fs';
import path from 'node:path';
import { paths, readJson } from '../cli.ts';
import { load, workflowHash } from '../contract.ts';
import { WorkflowError } from '../errors.ts';
import { outputProblem, receiptProblem, sourceSnapshot } from '../receipts.ts';
import { readState } from '../state.ts';

function verdict(
  location: Record<string, string>,
  contract: Record<string, unknown>,
  check: Record<string, unknown>,
  currentDigest: string | null,
): [string, string] {
  const receiptPath = path.join(location['receipts'], `${check['id']}.json`);
  if (!fs.existsSync(receiptPath) || !fs.statSync(receiptPath).isFile()) {
    return ['MISSING', 'no receipt recorded'];
  }
  if (currentDigest === null) {
    return ['UNKNOWN', 'source digest unavailable (not a committed git tree)'];
  }
  const receipt = readJson(receiptPath);
  let problem = receiptProblem(receipt, contract, check, currentDigest);
  if (problem === null) {
    problem = outputProblem(receipt, check, path.join(location['run'], 'output'));
  }
  if (problem !== null) {
    return ['STALE', problem];
  }
  return [
    'CURRENT',
    `passed, tests=${receipt['discoveredTests']}, at ${receipt['recordedAt']}`,
  ];
}

export function render(root: string): string {
  const location = paths(root);
  const contract = load(location['contract']);
  const state = readState(location['state']);
  const recorded = state === null ? 'planned' : (state['status'] as string);

  let currentDigest: string | null = null;
  if (state !== null) {
    try {
      currentDigest = sourceSnapshot(root)['digest'] as string;
    } catch {
      currentDigest = null;
    }
  }

  const verdicts: Array<[string, string, string]> = (
    contract['checks'] as Record<string, unknown>[]
  ).map(c => {
    const [tag, detail] = verdict(location, contract, c, currentDigest);
    return [c['id'] as string, tag, detail];
  });

  const stale = verdicts.some(([, tag]) => tag !== 'CURRENT');
  const effective = recorded === 'verified' && stale ? 'implementing' : recorded;

  const out: string[] = [`# Workflow handoff — ${contract['id']}`, ''];
  const label =
    `**${effective}**` + (effective !== recorded ? ` (recorded ${recorded})` : '');
  out.push(`- State: ${label}`);
  out.push(`- Objective: ${contract['objective']}`);
  if (state !== null && state['blockReason']) {
    out.push(`- Blocked: ${state['blockReason']}`);
  }

  out.push('', '## Receipts');
  for (const [cid, tag, detail] of verdicts) {
    out.push(`- \`${cid}\`: **${tag}** — ${detail}`);
  }

  out.push('', '## Closet');
  out.push(`- Workflow id: \`${contract['id']}\``);
  out.push(`- Workflow hash: \`${workflowHash(contract)}\``);
  const scopeStr = (contract['scope'] as string[]).map(p => `\`${p}\``).join(', ') || '(none)';
  out.push(`- Scope: ${scopeStr}`);
  const exclusions = contract['exclusions'] as string[];
  if (exclusions.length > 0) {
    out.push(`- Exclusions: ${exclusions.map(p => `\`${p}\``).join(', ')}`);
  }
  const artifactsStr =
    (contract['artifacts'] as string[]).map(p => `\`${p}\``).join(', ') || '(none)';
  out.push(`- Artifacts: ${artifactsStr}`);

  out.push('', '## Rebuild to verified');
  if (effective === 'verified') {
    out.push('- Already verified with current receipts. No action needed.');
  } else {
    const steps: string[] = [];
    if (effective === 'planned' || effective === 'blocked') {
      steps.push('`begin` (required before any check)');
    }
    for (const [cid, tag] of verdicts) {
      if (tag !== 'CURRENT') {
        steps.push(`\`run-check ${cid}\` (currently ${tag})`);
      }
    }
    steps.push('`verify`');
    for (let i = 0; i < steps.length; i++) {
      out.push(`${i + 1}. ${steps[i]}`);
    }
  }
  out.push('');
  return out.join('\n');
}

export function main(argv: string[] = process.argv.slice(2)): void {
  const args = argv.slice();
  const rootIdx = args.indexOf('--root');
  let rootPath = '.';
  if (rootIdx !== -1) {
    rootPath = args[rootIdx + 1] ?? '.';
  }
  try {
    process.stdout.write(render(path.resolve(rootPath)));
  } catch (e) {
    if (e instanceof WorkflowError) {
      process.stderr.write(`render_handoff error: ${e.message}\n`);
      process.exit(1);
    }
    throw e;
  }
}
